### ISSUE TITLE
Offscreen rendering mode shows unhideable dummy window

### ISSUE THREAD

**Issue by vulture** [electron/electron#8392]
* Electron version: 1.4.12
* Operating system: Win 8.1

### Expected behavior

We should be able to create an offscreen renderer window without showing any dummy window. Additionally, user and js performing operations on this dummy window (e.g. .hide()) should not resize the window. It should leave the offscreen renderer alone at its specified resolution.

### Actual behavior

Various actions on the dummy window affect the offscreen renderer as well.

**Comment by gerhardberger** [comment]
By passing `show: false` to the `BrowserWindow` constructor the dummy window should not be visible. In your case is it still showing?

The dummy window is necessary in offscreen mode, because Chromium needs a window handle.

**Comment by vulture** [comment]
So after some more investigation, there are a few issues happening:
1. Even if I pass "show: false", there is still a brief flash of the dummy window appearing and then hiding, which looks unprofessional to the end user. I can make this smaller by passing in width: 0 and height: 0, but there is still a minimum size where a white rectangle shows up for a brief period. Setting x: and y: to negative values does not hide this offscreen.
2. The offscreen renderer is supposed to be a frameless window by default (see: https://github.com/electron/electron/blob/master/docs/tutorial/offscreen-rendering.md). However, the dummy window created does not take this default and is a framed window (presumably so that you can drag it around on screen with a titlebar etc). This is problematic because when calling SetSize() or SetContentSize(), the window can get re-resized to a value you didn't choose. For example, if I try to resize the window to 1920x1080, it instead gets sized to 1910x1027 or 1920x1039 (presumably to accommodate the framed window borders). So, it is probably using the dummy window for all this. I need to explicitly specify "frame: false" for "fullscreen" window sizes to work.
3. If I call .show() on the window, I actually do get a 1920x1080 window after all this, but then the dummy window shows up again.

Currently the best set of parameters is something like
```
  win = new BrowserWindow({
    show: false,
    frame: false,
    webPreferences: {
      offscreen: true,
      transparent: true,
    },
  })
```
and then I need to be sure not to call .show().

In conclusion:
1. "frame: false" is necessary so that the dummy window does not get a frame, because the [offscreen renderer content] size can/will be wrong if you want a resolution near the full screen.
2. This still flashes the dummy window briefly when the BrowserWindow gets created(?).

**Comment by gerhardberger** [comment]
@vulture 

1. Since 1.4.14 offscreen windows are frameless by default: #8250. You should upgrade to this version.
2. On macOS I've never experienced this flashing nor people I know using it on Windows. Could you provide an example program that gives you this flashing?

**Comment by vulture** [comment]
After some further investigation (also updated to 1.4.14), I found that my understanding was slightly off. I was mirroring the OSR frame over top my desktop so that I could see it. As it turns out there is no window that flashes. It is actually the OSR framebuffer that (incorrectly?) has a white rectangle the size of the dummy window(?), and since I was just drawing that on my desktop, it seemed like it must actually be drawing a dummy window itself on the desktop. So the problem is actually in the OSR framebuffer. There may actually be nothing wrong with the dummy window at all.

Here is an imgur album of the saved frames out: http://imgur.com/a/IFUCo
The damage_rect values are x,y=0,0 width,height=1920,1080 for all of these frames.

Because I'm using transparent: true, maybe there's an issue somewhere that does not respect transparent and instead clears some rect on the OSR framebuffer using white instead of transparent? The white value is 0xffffffff (so, alpha really is 0xff too, not 0x00). The flashing does not happen every time, but it does usually happen for me. Sometimes instead, the first frame is correctly all black, and then the second or third frame has the white rectangle, and then it disappears again. I am calling window()->SetContentSize to set the OSR size to 1920x1080, in a js callback, before the first OnPaint. Not sure if that could be causing any funny issues or not, but thought I'd throw that in there.

```
const {app, BrowserWindow} = require('electron')
app.disableHardwareAcceleration()
let win
app.once('ready', () => {
  win = new BrowserWindow({
    show: false,
    frame: false,
    webPreferences: {
      offscreen: true,
      transparent: true,
    },
  })
  win.loadURL('...whatever...')
  win.webContents.setFrameRate(30);
  win.webContents.on('paint', (event, dirty, image) => {  ... same bmp ... })
})

```

**Comment by gerhardberger** [comment]
It is up to the Chromium renderer to generate the frames that are emitted in the `paint` callback. There are usually frames where the website not fully rendered, but I think it generates these frames to show content as quickly as possible.

I was able to avoid these initial white frames by setting a flag in BrowserWindow's `ready-to-show` event, to only handle paint events when that event happened.

``` js
const {app, BrowserWindow} = require('electron')
app.disableHardwareAcceleration()
let win
app.once('ready', () => {
  win = new BrowserWindow({
    show: false,
    frame: false,
    webPreferences: {
      offscreen: true,
      transparent: true
    }
  })
  win.loadURL('https://www.github.com')
  win.webContents.setFrameRate(30)

  let ready = false
  win.on('ready-to-show', () => (ready = true))
  win.webContents.on('paint', (event, dirty, image) => {
    if (!ready) return
    // handle frame
  })
})
```

**Comment by vulture** [comment]
I tried this instead:
```
        mainWindow.on('ready-to-show', () => {
            mainWindow.webContents.on('paint', (event, dirty, image) => { ... });
        });
```
which did not work. Due to the way I handle paint internally I can't actually use a js var 'ready' like that. If there is not a real solution in Electron itself and Chromium is to blame, I have a different hacky solution I guess I can use, but it is not optimal.

**Comment by vulture** [comment]
Okay turns out this seems to work pretty well. Due to how I was handling the paint event I actually needed to emulate the ready-to-show a different way (internal to Electron's code). I haven't been able to repro the issue since using this fix, and it seems to work normally otherwise, so I would consider this a working workaround. Ideally OnPaint wouldn't get called in the first place without a valid frame, but, I'm happy enough with this fix. :P

**Comment by gerhardberger** [comment]
Nice, happy to see you got it working!

**Comment by JerryI** [comment]
I would assign this issue to `resolved`.
Just checked on the latest electron with `offscreen: true, ...show:false`. Nothing pops up
